go/pools/resource_pool.go

```diff
@@ -205,34 +222,110 @@ func (rp *ResourcePool) reopen() {
-func (rp *ResourcePool) Get(ctx context.Context) (resource Resource, err error) {
-	return rp.get(ctx)
-}
-
-func (rp *ResourcePool) get(ctx context.Context) (resource Resource, err error) {
+func (rp *ResourcePool) Get(ctx context.Context, settings []string) (resource Resource, err error) {
 	// If ctx has already expired, avoid racing with rp's resource channel.
 	if ctx.Err() != nil {
 		return nil, ErrCtxTimeout
 	}
+	if len(settings) == 0 {
+		return rp.get(ctx)
+	}
+	return rp.getWithSettings(ctx, settings)
+}
 
+func (rp *ResourcePool) get(ctx context.Context) (resource Resource, err error) {
 	// Fetch
 	var wrapper resourceWrapper
 	var ok bool
+	// If we put both the channel together, then, go select can read from any channel
+	// this way we guarantee it will try to read from the channel we intended to read it from first
+	// and then try to read from next best available resource.
 	select {
+	// check normal resources first
 	case wrapper, ok = <-rp.resources:
 	default:
-		startTime := time.Now()
 		select {
+		// then checking setting resources
+		case wrapper, ok = <-rp.settingResources:
+		default:
+			// now waiting
+			startTime := time.Now()
+			select {
+			case wrapper, ok = <-rp.resources:
+			case wrapper, ok = <-rp.settingResources:
+			case <-ctx.Done():
+				return nil, ErrTimeout
+			}
+			rp.recordWait(startTime)
+		}
+	}
+	if !ok {
+		return nil, ErrClosed
+	}
+
+	// if the resource has settings applied, we will close it and return a new one
+	if wrapper.resource != nil && wrapper.resource.SettingHash() != 0 {
+		wrapper.resource.Close()
+		wrapper.resource = nil
+		rp.active.Add(-1)
+	}
+
+	// Unwrap
+	if wrapper.resource == nil {
+		wrapper.resource, err = rp.factory(ctx)
+		if err != nil {
+			rp.resources <- resourceWrapper{}
+			return nil, err
+		}
+		rp.active.Add(1)
+	}
+	if rp.available.Add(-1) <= 0 {
+		rp.exhausted.Add(1)
+	}
+	rp.inUse.Add(1)
+	return wrapper.resource, err
+}
+
+func (rp *ResourcePool) getWithSettings(ctx context.Context, settings []string) (Resource, error) {
+	var wrapper resourceWrapper
+	var ok bool
+
+	settingHash, err := hash(settings)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fetch
+	select {
+	// check setting resources first
+	case wrapper, ok = <-rp.settingResources:
+	default:
+		select {
+		// then, check normal resources
 		case wrapper, ok = <-rp.resources:
-		case <-ctx.Done():
-			return nil, ErrTimeout
+		default:
+			// now waiting
+			startTime := time.Now()
+			select {
+			case wrapper, ok = <-rp.settingResources:
+			case wrapper, ok = <-rp.resources:
+			case <-ctx.Done():
+				return nil, ErrTimeout
+			}
+			rp.recordWait(startTime)
 		}
-		rp.recordWait(startTime)
 	}
 	if !ok {
 		return nil, ErrClosed
 	}
 
+	// Checking setting hash id, if it is different, we will close the resource and return a new one later in unwrap
+	if wrapper.resource != nil && wrapper.resource.SettingHash() > 0 && wrapper.resource.SettingHash() != settingHash {
```

Comment:
shouldn't we check if the settings are actually equal here? just checking the hash could lead to situations where to different settings yield the same hash code, right?

Reply:
in the next iteration will get rid hash itself and the settings will become []string to just string. so comparison will be simpler.